login_server/player/state/authenticated_state.py

```diff
@@ -127,17 +132,30 @@ def handle_server_join_first_step(self, request):
         else:
             game_server = self.player.login_server.find_server_by_id(server_field.value)
 
+            if game_server.game_setting_mode != self.player.player_settings.game_setting_mode:
+                # Cannot join a goty server in ootb mode or vice versa
+                self._send_private_msg_from_server(self.player, 'You are in %s mode; you cannot join a %s mode server' %
+                                                   (self.player.player_settings.game_setting_mode,
+                                                    game_server.game_setting_mode))
+                return
+
+            # Also disallow joining a non-ootb server if the player is not known to be modded
+            if game_server.game_setting_mode != 'ootb' and not self.player.is_modded:
+                self._send_private_msg_from_server(self.player, 'You cannot join a %s server without TAMods' %
```

Comment:
Put this before the earlier check, because players without TAMods can't switch mode (right?)

Reply:
Done